pkg/plugins/runtime/k8s/webhooks/validation.go

```diff
@@ -103,40 +109,109 @@ func (h *validatingHandler) decode(req admission.Request) (core_model.Resource,
+			return nil, nil, err
+		}
 	}
+
 	if err := h.converter.ToCoreResource(k8sObj, coreRes); err != nil {
 		return nil, nil, err
 	}
 	return coreRes, k8sObj, nil
 }
 
 // Note that this func does not validate ConfigMap and Secret since this webhook does not support those
-func (h *validatingHandler) isOperationAllowed(resType core_model.ResourceType, userInfo authenticationv1.UserInfo) admission.Response {
-	if slices.Contains(h.allowedUsers, userInfo.Username) {
-		// Assume this means one of the following:
-		// - sync from another zone (rt.Config().Runtime.Kubernetes.ServiceAccountName))
-		// - GC cleanup resources due to OwnerRef. ("system:serviceaccount:kube-system:generic-garbage-collector")
-		// - storageversionmigratior
-		// Not security; protecting user from self.
+func (h *validatingHandler) isOperationAllowed(userInfo authenticationv1.UserInfo, r core_model.Resource) admission.Response {
+	if h.isPrivilegedUser(userInfo) {
 		return admission.Allowed("")
 	}
 
-	descriptor, err := h.coreRegistry.DescriptorFor(resType)
-	if err != nil {
-		return syncErrorResponse(resType, h.mode)
-	}
-	if descriptor.KDSFlags == core_model.KDSDisabledFlag {
-		return admission.Allowed("")
+	if !h.isResourceTypeAllowed(r.Descriptor()) {
+		return resourceTypeIsNotAllowedResponse(r.Descriptor().Name, h.mode)
 	}
-	if (h.mode == core.Global && !descriptor.KDSFlags.Has(core_model.AllowedOnGlobalSelector)) || (h.federatedZone && !descriptor.KDSFlags.Has(core_model.AllowedOnZoneSelector)) {
-		return syncErrorResponse(resType, h.mode)
+
+	if !h.isResourceAllowed(r) {
+		return resourceIsNotAllowedResponse()
 	}
+
 	return admission.Allowed("")
 }
 
-func syncErrorResponse(resType core_model.ResourceType, cpMode core.CpMode) admission.Response {
+func (h *validatingHandler) isPrivilegedUser(userInfo authenticationv1.UserInfo) bool {
+	// Assume this means one of the following:
+	// - sync from another zone (rt.Config().Runtime.Kubernetes.ServiceAccountName)
+	// - GC cleanup resources due to OwnerRef. ("system:serviceaccount:kube-system:generic-garbage-collector")
+	// - storageversionmigratior
+	// Not security; protecting user from self.
+	return slices.Contains(h.allowedUsers, userInfo.Username)
+}
+
+func (h *validatingHandler) isResourceTypeAllowed(d core_model.ResourceTypeDescriptor) bool {
+	if d.KDSFlags == core_model.KDSDisabledFlag {
+		return true
+	}
+	if h.mode == core.Global && !d.KDSFlags.Has(core_model.AllowedOnGlobalSelector) {
+		return false
+	}
+	if h.federatedZone && !d.KDSFlags.Has(core_model.AllowedOnZoneSelector) {
+		return false
+	}
+	return true
+}
+
+func (h *validatingHandler) isResourceAllowed(r core_model.Resource) bool {
+	if !h.federatedZone || !r.Descriptor().IsPluginOriginated {
+		return true
+	}
+	if !h.disableOriginLabelValidation {
+		if r.GetMeta().GetLabels() == nil || r.GetMeta().GetLabels()[mesh_proto.ResourceOriginLabel] != string(mesh_proto.ZoneResourceOrigin) {
```

Comment:
nit: shouldn't we use this `ResourceOrigin()`?

Reply:
ah missed that place! thanks